lib/resourcebuilder/batch.go

```diff
@@ -46,15 +47,36 @@ func checkJobHealth(ctx context.Context, client batchclientv1.JobsGetter, job *b
 	}
 
 	// Since we have filled in "activeDeadlineSeconds",
-	// the Job will 'Active == 0' if and only if it exceeds the deadline.
+	// the Job will 'Active == 0' if and only if it exceeds the deadline or if the update image could not be pulled.
 	// Failed jobs will be recreated in the next run.
-	if j.Status.Active == 0 && j.Status.Failed > 0 {
-		reason := "DeadlineExceeded"
-		message := "Job was active longer than specified deadline"
-		if len(j.Status.Conditions) > 0 {
-			reason, message = j.Status.Conditions[0].Reason, j.Status.Conditions[0].Message
+	if j.Status.Active == 0 {
+		klog.V(4).Infof("No active pods for job %s in namespace %s", job.Name, job.Namespace)
+		failed, reason, message := hasJobFailed(job)
+		// If there is more than one failed job pod then get the cause for failure
+		if j.Status.Failed > 0 {
+			failureReason := "DeadlineExceeded"
+			failureMessage := "Job was active longer than specified deadline"
+			if failed {
+				failureReason, failureMessage = reason, message
+			}
+			return false, fmt.Errorf("deadline exceeded, failureReason: %q, message: %q", failureReason, failureMessage)
+		}
+
+		// When the update image cannot be pulled then the pod is not marked as Failed, but the status condition is set
+		// after the job deadline is exceeded.
+		if failed && reason == "DeadlineExceeded" {
```

Comment:
I don't think we need the `reason` check here.  Any reason out of `hasJobFailed`'s `Failed=True` condition should be fine for complaining about, right?

Reply:
I'm not sure. Ideally all the conditions should be present and the `Failed` condition should only be set when the Job has actually failed. But the conditions on the `Job` don't seem very consistent compared to other types of resources. I'm trying to keep the logic the same as it was before and not report a failure when there is chance that the job would be successful at some point in the future.

Reply:
can we at least log the `Failed=True` reason/message we're going to ignore, then?  Something like:

```go
if failed {
  if reason == "DeadlineExceeded" {
    return false, fmt.Errorf("deadline exceeded, reason: %q, message: %q", reason, message)
  } else {
    klog.V(4).Infof("Ignoring job %s in namespace %s setting Failed=True with %s: %s", job.Name, job.Namespace, reason, message)
   }
 }
 
 return false, nil
 ```

Reply:
Done